src/Compilers/Core/Portable/DiagnosticAnalyzer/CompilationWithAnalyzers.cs

```diff
@@ -633,28 +633,15 @@ public Task<ImmutableArray<Diagnostic>> GetAllDiagnosticsAsync()
                 // Get instance of analyzer driver from the driver pool.
                 AnalyzerDriver driver = _driverPool.Allocate();
 
-                try
-                {
-                    // Start the initialization task, if required.
-                    if (driver.WhenInitializedTask == null)
-                    {
-                        driver.Initialize(_compilation, _analysisOptions, _compilationData, categorizeDiagnostics: true, cancellationToken: cancellationToken);
-                    }
-
-                    // Wait for driver initialization to complete: this executes the Initialize and CompilationStartActions to compute all registered actions per-analyzer.
-                    await driver.WhenInitializedTask.ConfigureAwait(false);
-                }
-                finally
+                // Start the initialization task, if required.
+                if (driver.WhenInitializedTask == null)
                 {
-                    if (driver.WhenInitializedTask.IsCanceled)
-                    {
-                        // If the initialization task was cancelled, we retry again with our own cancellation token.
-                        // This can happen if the task that started the initialization was cancelled by the callee, and the new request picked up this driver instance.
-                        _driverPool.ForgetTrackedObject(driver);
-                        driver = await GetAnalyzerDriverAsync(cancellationToken).ConfigureAwait(false);
-                    }
+                    driver.Initialize(_compilation, _analysisOptions, _compilationData, categorizeDiagnostics: true, cancellationToken: cancellationToken);
                 }
 
+                // Wait for driver initialization to complete: this executes the Initialize and CompilationStartActions to compute all registered actions per-analyzer.
+                await driver.WhenInitializedTask.ConfigureAwait(false);
+
                 return driver;
             }
             catch (Exception e) when (FatalError.ReportUnlessCanceled(e))
```

Comment:
The callers of GetAnalyzerDriverAsync are expected to invoke it in a try clause, with a FreeDriver invocation in the finally.